notifications.go

```diff
@@ -0,0 +1,271 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/go-telegram-bot-api/telegram-bot-api"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"strconv"
+	"sync"
+)
+
+const (
+	notificationsDB = "notifications.json"
+)
+
+type notifications struct {
+	sync.RWMutex
+	Users map[string]string `json:"users"`
+}
+
+var (
+	userNotifications = notifications{Users: make(map[string]string)}
+)
+
+// notificationHandler enables/disables user notifications.
+func (x *opBot) notificationHandler(update tgbotapi.Update) error {
+	uidstr := fmt.Sprintf("%d", update.Message.From.ID)
+
+	if err := toggleNotifications(uidstr, update.Message.From.UserName); err != nil {
+		text := fmt.Sprintf("%s (%s)", T("notification_fail"), notificationStatus(uidstr))
+		x.sendReply(update, text)
+		log.Println(err)
+		return nil
+	}
+
+	text := fmt.Sprintf("%s (%s)", T("notification_success"), notificationStatus(uidstr))
+	x.sendReply(update, text)
+	return nil
+}
+
+// notificationStatus returns a string with the enabled/disabled status of notifications
+// for the given user.
+func notificationStatus(user string) string {
+	if notificationsEnabled(user) {
+		return T("notifications_enabled")
+	}
+	return T("notifications_disabled")
+}
+
+// notificationsEnabled returns a boolean indicating whether
+// the user passed as argument has enabled notifications.
+func notificationsEnabled(user string) bool {
+	userNotifications.RLock()
+	defer userNotifications.RUnlock()
+
+	_, ok := userNotifications.Users[user]
+	return ok
+}
+
+// toggleNotifications is going to change the current notification
+// settings for the specific user, and save the resulting config
+// to the notificationsDB file.
+func toggleNotifications(userid, username string) error {
+	userNotifications.Lock()
+	defer userNotifications.Unlock()
+
+	savedUserName, ok := userNotifications.Users[userid]
+	if ok {
+		delete(userNotifications.Users, userid)
+		if len(savedUserName) > 0 {
+			delete(userNotifications.Users, savedUserName)
+		}
+	} else {
+		userNotifications.Users[userid] = username
+		if len(username) > 0 {
+			userNotifications.Users[username] = userid
+		}
+	}
+
+	err := saveNotifications()
+	if err != nil {
+		// Let's undo what we did here, as we were unable to
+		// persist the settings in the notifications database.
+		if ok {
+			userNotifications.Users[userid] = savedUserName
+			if len(username) > 0 {
+				userNotifications.Users[username] = userid
+			}
+		} else {
+			delete(userNotifications.Users, userid)
+			if len(savedUserName) > 0 {
+				delete(userNotifications.Users, savedUserName)
+			}
+		}
+	}
+
+	return err
+}
+
+// saves notifications info to notificationsDB file.
+func saveNotifications() error {
+	buf, err := json.Marshal(userNotifications.Users)
+	if err != nil {
+		return err
+	}
+
+	datadir, err := createDataDirIfDoesntExist()
+	if err != nil {
+		return err
+	}
+
+	f := filepath.Join(datadir, notificationsDB)
+	return ioutil.WriteFile(f, buf, 0644)
```

Comment:
It's a really  good idea to save into a tmp file and then move the tmp file on top of the original file. This guarantees an atomic operation.